tests/bsim/bluetooth/host/att/sequential/tester/src/main.c

```diff
@@ -0,0 +1,715 @@
+	struct net_buf *buf;
+	uint64_t mask = 0U;
+
+	/* The two commands have the same length/params */
+	buf = bt_hci_cmd_create(opcode, sizeof(*cp_mask));
+	__ASSERT_NO_MSG(buf);
+
+	/* Forward all events */
+	cp_mask = net_buf_add(buf, sizeof(*cp_mask));
+	mask = UINT64_MAX;
+	sys_put_le64(mask, cp_mask->events);
+
+	send_cmd(opcode, buf, NULL);
+}
+
+static void set_random_address(void)
+{
+	struct net_buf *buf;
+	bt_addr_le_t addr = {BT_ADDR_LE_RANDOM, {{0x0A, 0x89, 0x67, 0x45, 0x23, 0xC1}}};
+
+	LOG_DBG("%s", bt_addr_str(&addr.a));
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_RANDOM_ADDRESS, sizeof(addr.a));
+	__ASSERT_NO_MSG(buf);
+
+	net_buf_add_mem(buf, &addr.a, sizeof(addr.a));
+	send_cmd(BT_HCI_OP_LE_SET_RANDOM_ADDRESS, buf, NULL);
+}
+
+void start_adv(uint16_t interval)
+{
+	struct bt_hci_cp_le_set_adv_param set_param;
+	struct net_buf *buf;
+
+	(void)memset(&set_param, 0, sizeof(set_param));
+
+	set_param.min_interval = sys_cpu_to_le16(interval);
+	set_param.max_interval = sys_cpu_to_le16(interval);
+	set_param.channel_map = 0x07;
+	set_param.filter_policy = BT_LE_ADV_FP_NO_FILTER;
+	set_param.type = BT_HCI_ADV_IND;
+	set_param.own_addr_type = 0x01; /* random */
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_ADV_PARAM, sizeof(set_param));
+	__ASSERT_NO_MSG(buf);
+	net_buf_add_mem(buf, &set_param, sizeof(set_param));
+
+	send_cmd(BT_HCI_OP_LE_SET_ADV_PARAM, buf, NULL);
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_ADV_ENABLE, 1);
+	__ASSERT_NO_MSG(buf);
+
+	net_buf_add_u8(buf, BT_HCI_LE_ADV_ENABLE);
+	send_cmd(BT_HCI_OP_LE_SET_ADV_ENABLE, buf, NULL);
+}
+
+NET_BUF_POOL_DEFINE(acl_tx_pool, 5, BT_L2CAP_BUF_SIZE(200), 8, NULL);
+
+struct net_buf *alloc_l2cap_pdu(void)
+{
+	struct net_buf *buf;
+	uint16_t reserve;
+
+	buf = net_buf_alloc(&acl_tx_pool, K_FOREVER);
+	__ASSERT_NO_MSG(buf);
+
+	reserve = sizeof(struct bt_l2cap_hdr);
+	reserve += sizeof(struct bt_hci_acl_hdr) + BT_BUF_RESERVE;
+
+	net_buf_reserve(buf, reserve);
+
+	return buf;
+}
+
+static int send_acl(struct net_buf *buf)
+{
+	struct bt_hci_acl_hdr *hdr;
+	uint8_t flags = BT_ACL_START_NO_FLUSH;
+
+	hdr = net_buf_push(buf, sizeof(*hdr));
+	hdr->handle = sys_cpu_to_le16(bt_acl_handle_pack(conn_handle, flags));
+	hdr->len = sys_cpu_to_le16(buf->len - sizeof(*hdr));
+
+	bt_buf_set_type(buf, BT_BUF_ACL_OUT);
+
+	k_sem_take(&acl_pkts, K_FOREVER);
+
+	return bt_send(buf);
+}
+
+static void send_l2cap_packet(struct net_buf *buf, uint16_t cid)
+{
+	struct bt_l2cap_hdr *hdr;
+
+	hdr = net_buf_push(buf, sizeof(*hdr));
+	hdr->len = sys_cpu_to_le16(buf->len - sizeof(*hdr));
+	hdr->cid = sys_cpu_to_le16(cid);
+
+	/* Always entire packets, no HCI fragmentation */
+	__ASSERT_NO_MSG(buf->len <= CONFIG_BT_BUF_ACL_TX_SIZE);
+
+	send_acl(buf);
+}
+
+static void gatt_write(bool req)
+{
+	static uint8_t data[] = "write";
+	uint16_t handle = server_write_handle;
+	struct net_buf *buf = alloc_l2cap_pdu();
+
+	net_buf_add_u8(buf, req ?
+		       BT_ATT_OP_WRITE_REQ : BT_ATT_OP_WRITE_CMD);
```

Comment:
When I read the `gatt_write(false)`, and `gatt_write(true)` lines below, I thought that it may be more readable if this was `gatt_write(BT_ATT_OP_WRITE_REQ)` and `gatt_write(BT_ATT_OP_WRITE_CMD)`. What do you think?

Reply:
That's better, yeah